src/Game/AITools.h

```diff
@@ -0,0 +1,95 @@
+
+//
+// Created by tim on 14.06.19.
+//
+
+#ifndef KI_AITOOLS_H
+#define KI_AITOOLS_H
+
+#include <memory>
+#include <optional>
+#include <deque>
+#include <list>
+#include <functional>
+#include <set>
+
+namespace aiTools{
+    template <typename T>
+    class SearchNode {
+    public:
+        SearchNode(T state, std::optional<std::shared_ptr<SearchNode<T>>> parent, int pathCost) :
+            state(state), parent(parent), pathCost(pathCost) {}
+
+        T state;
+        std::optional<std::shared_ptr<SearchNode<T>>> parent;
+        int pathCost = 0;
+        bool operator==(const SearchNode<T> &other) const;
+
+    };
+
+
+    template<typename T>
+    bool SearchNode<T>::operator==(const SearchNode<T> &other) const {
+        return this->state == other.state;
+    }
+
+
+    /**
+     * A*algorithm implemented as graph search
+     * @tparam T Type of the SearchNode used
+     * @param Start initial state
+     * @param Destination goal state
+     * @param expand Function used for expanding nodes
+     * @param f Function used for evaluating a node. A high value corresponds to low utility
+     * @return The goal node with information about its parent or nothing if problem is impossible
+     */
+    template <typename T>
+    auto aStarSearch(SearchNode<T> &start, SearchNode<T> &destination,
+            std::function<std::vector<SearchNode<T>>(const SearchNode<T>&)> expand,
+            std::function<int(const SearchNode<T>&)> f) -> std::optional<SearchNode<T>>{
+        std::deque<SearchNode<T>> visited;
+        std::list<SearchNode<T>> fringe = {start};
+        std::optional<SearchNode<T>> ret = std::nullopt;
+        while(!fringe.empty()){
+            const auto currentNode = *fringe.begin();
+            fringe.pop_front();
+            if(currentNode == destination){
+                ret = currentNode;
+                break;
+            }
+
+            bool nodeVisited = false;
+            for(const auto &node : visited){
+                if(currentNode == node){
+                    nodeVisited = true;
+                    break;
+                }
+            }
+
+            if(nodeVisited){
+                continue;
+            }
+
+            visited.emplace_back(currentNode);
+            for(const auto &newNode : expand(currentNode)){
+                bool inserted = false;
+                for(auto it = fringe.begin(); it != fringe.end(); it++){
```

Comment:
```suggestion
                for(auto it = fringe.begin(); it != fringe.end(); ++it){
```
Ist allgemein in jeder for-Schleife theoretisch performanter das als pre-increment (statt post) zu machen (`add` + wert zurückgeben vs. kopie vom alten Wert machen und dann `add`). Bei trivialen typen wie `int` wird der compiler das optimieren, bei nicht trivialen typen mit überladenen Operatoren geht das nicht, da nicht garantiert werden kann, dass `++it` und `it++` das gleiche machen